<?xml version="1.0"?>
<launch>

    <!-- Associate the robot description parameter with the robot's URDF -->
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find pose_estimation)/urdf/ur5_cam.xacro" />
    
    <!-- Associate the table description parameter with the table's URDF -->
    <param name="table_description" command="$(find xacro)/xacro --inorder $(find pose_estimation)/urdf/object_table.urdf.xacro" />

    <!-- Load controller configurations -->
    <rosparam file="$(find pose_estimation)/config/ur5_controllers.yaml" command="load" />

    <!-- Set the robot's initial position in the world -->
    <arg name="x" default="0" />
    <arg name="y" default="0" />
    <arg name="z" default="0.5" />
    
    <!-- Launch Gazebo world in paused mode -->
    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="use_sim_time" value="true"/>
        <arg name="debug" value="false"/>
        <arg name="gui" value="true"/>
        <arg name="paused" value="true" />
    </include>   

    <!-- Spawn the robot in Gazebo -->
    <node name="spawn_the_robot" pkg="gazebo_ros" type="spawn_model" output="screen" 
          args="-urdf -param robot_description -model ur5_robot -x $(arg x) -y $(arg y) -z $(arg z)" />

    <!-- Spawn the table in Gazebo -->
    <node name="spawn_the_table" pkg="gazebo_ros" type="spawn_model" output="screen"
          args="-urdf -param table_description -model my_table -x 1.3 -y 0 -z $(arg z)" />

    <!-- Spawn additional objects if any -->
    <group ns="spawn_objects">
        <rosparam file="$(find pose_estimation)/config/spawn_objects.yaml"/>
    </group>
    <node name="spawn_objects_node" pkg="pose_estimation" type="spawn_objects.py" output="screen"/>
    
     <!-- Load the node controller manager-->
    <node name="controller_spawner" pkg="controller_manager" type="spawner"
            output="screen" respawn="false" ns="/ur5"
            args="joint_state_controller
                eff_joint_traj_controller
                --timeout 60 ">
    </node>

</launch>
